pkg/rpcmiddleware/validator/validator.go

```diff
@@ -60,17 +86,53 @@ func convertError(err error) error {
 	return grpc.Errorf(codes.InvalidArgument, err.Error())
 }
 
-var errNoValidator = errors.DefineUnimplemented("no_validator", "validator not defined")
+var (
+	errNoValidator = errors.DefineUnimplemented("no_validator", "validator not defined")
+	errNonZeroPath = errors.DefineInvalidArgument("non_zero_path", "path '{path}' is not zero")
+)
 
 func validateMessage(ctx context.Context, fullMethod string, msg interface{}) error {
 	if v, ok := msg.(interface {
 		GetFieldMask() types.FieldMask
 	}); ok {
 		region := trace.StartRegion(ctx, "validate field mask")
-		if forbiddenPaths := forbiddenPaths(v.GetFieldMask().Paths, allowedFieldMaskPaths[fullMethod]); len(forbiddenPaths) > 0 {
+
+		paths := v.GetFieldMask().Paths
+		if forbiddenPaths := forbiddenPaths(paths, allowedFieldMaskPaths[fullMethod]); len(forbiddenPaths) > 0 {
 			region.End()
 			return errForbiddenFieldMaskPaths.WithAttributes("forbidden_paths", forbiddenPaths)
 		}
+
+		disallowedPaths := disallowedFieldMaskPaths[fullMethod]
+		var isZero func(string) bool
+		for _, p := range paths {
+			prefix := p + "."
+			i := sort.Search(len(disallowedPaths), func(i int) bool {
+				return strings.HasPrefix(disallowedPaths[i], prefix)
+			})
+			if i == len(disallowedPaths) || !strings.HasPrefix(disallowedPaths[i], prefix) {
+				continue
+			}
+			tail := disallowedPaths[i+1:]
+			for _, sp := range disallowedPaths[i : i+1+sort.Search(len(tail), func(j int) bool {
+				return !strings.HasPrefix(tail[j], prefix)
+			})] {
+				if isZero == nil {
+					if v, ok := msg.(interface {
+						FieldIsZero(string) bool
+					}); ok {
+						isZero = v.FieldIsZero
+					} else {
+						// TODO: Use reflection to validate a subfield by path
+						break
```

Comment:
So if `msg` is of type `EndDevice`, and `FieldIsZero` is implemented on `MACState`, it doesn't get called unless `EndDevice` implements some umbrella `FieldIsZero`? I think it makes sense to check the nearest `FieldIsZero` of `sp` here.

If we don't do this, this `FieldIsZero` support on `msg` should be done before the loop.

Reply:
> I think it makes sense to check the nearest FieldIsZero of sp here.

1. Since these methods will be auto-generated(see https://github.com/TheThingsNetwork/lorawan-stack/pull/3647#discussion_r553969216), all relevant messages will have this implemented
2. We do not have a generic way to access value of type `T` under arbitrary path `a.b.c`, so that is not possible unless we loop over all field tree in the struct using reflection and I don't think we should do that for all RPCs on each call.
3. `FieldIsZero` check is done inside the loop to avoid unnecessary type-assert on each RPC. This check is only performed when there's a path, which needs to be zero - there's no reason to check `FieldIsZero`, if we don't need to call it.